### PR TITLE
Use templates in RDKit coordinate generation

### PR DESCRIPTION
**Pull request by rachelnwalker**
#### What does this implement/fix? Explain your changes.

Draft PR for adding templates to RDKit native coordinate generation.

#### Any other comments?

These changes: https://gist.github.com/rachelnwalker/14475f9140a0afacae3502292d9b7097
Gist with templates: https://gist.github.com/rachelnwalker/c014eb8e83fb0f94cfd1f6daca6f8f18
Comparing rdkit and coordgen coordinates: https://gist.github.com/rachelnwalker/f3311fc807eea2fd262f995c020e516b

### CODE REVIEW

**Review by greglandrum — COMMENTED** [review]
Before I go further into this: what kind of feedback would be most useful to you at this stage?

Some general, very high-level, comments on what's here:
- I think the idea of allowing templates to be used makes a ton of sense and will really help with depictions.
- CXSMILES is compact and easy to produce, which is nice if we want to make it easy for people to add new templates. 
- We probably should add an option to allow people to disable the use of templates if they want.

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp [hunk 1]
```diff
@@ -341,6 +342,101 @@ int EmbeddedFrag::findNeighbor(
   return -1;
 }
 
+bool EmbeddedFrag::matchToTemplate(const RDKit::INT_VECT &ringSystemAtoms, unsigned int ring_count) {
+  CoordinateTemplates& coordinate_templates = CoordinateTemplates::getRingSystemTemplates();
+  if (!coordinate_templates.hasTemplateOfSize(ringSystemAtoms.size())) {
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/EmbeddedFrag.cpp:347]: It's probably worth a comment somewhere to discuss why we only use templates that 100% overlap with the ring system. Like, why not use a template that matches _some_ of the right system?

File: Code/GraphMol/Depictor/Templates.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+ private:
+  CoordinateTemplates() {
+    // load templates into m_templates map by atom count
+    std::ifstream cxsmiles_templates ("templates.smi");
+    std::string smiles;
+    while(std::getline(cxsmiles_templates, smiles)) {
+        auto mol = RDKit::SmilesToMol(smiles);
+        RDKit::MolOps::symmetrizeSSSR(*mol);
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/Templates.h:49]: were you maybe intending to turn off sanitization? otherwise I thought that sssr happens by default in the reader.

File: Code/GraphMol/Depictor/Templates.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if (m_templates.find(atom_count) == m_templates.end()) {
+            m_templates[atom_count] = {mol};
+        } else {
+            m_templates[atom_count].push_back(mol);
+        }
+    }
+    cxsmiles_templates.close();
+  }
+
+  CoordinateTemplates(const CoordinateTemplates&) = delete;
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/Templates.h:60]: cool, thanks for the noncopyable by default

File: Code/GraphMol/Depictor/templates.smi [hunk 4]
```diff
@@ -0,0 +1,78 @@
+c1cc2ccc1-c1ccc(cc1)-c1ccc(cc1)-c1ccc(cc1)-c1ccc(cc1)-c1ccc-2cc1 |(-1.23952,-2.13951,;-0.00687016,-1.42979,;-0.00517751,-0.007368,;-1.23623,0.705237,;-2.46888,-0.00448054,;-2.47057,-1.4269,;-3.45469,-3.12841,;-4.68654,-3.83952,;-4.68654,-5.26195,;-3.45479,-5.97316,;-2.22294,-5.26204,;-2.22285,-3.83962,;-2.47077,-7.67477,;-2.46927,-9.09719,;-1.23673,-9.8071,;-0.00567535,-9.0946,;-0.00716886,-7.67228,;-1.23972,-6.96236,;1.96009,-9.0948,;1.96188,-7.67248,;3.19452,-6.96276,;4.42548,-7.67536,;4.42379,-9.09778,;3.19114,-9.8075,;5.4096,-5.97385,;6.64145,-5.26274,;6.64145,-3.84032,;5.4096,-3.12911,;4.17775,-3.84032,;4.17775,-5.26274,;4.42568,-1.4275,;3.19452,-2.13991,;1.96198,-1.42999,;1.96058,-0.00756713,;3.19164,0.704839,;4.42418,-0.00507794,)|
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/templates.smi:1]: is the aromaticity important to these templates? I don't think that coordgen uses bond orders when matching templates, so we can probably drop that here. I also don't know if atomic number even matters!

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp [hunk 5]
```diff
@@ -341,6 +342,101 @@ int EmbeddedFrag::findNeighbor(
...
+    }
+    if (doneNbrs.empty()) {
+      d_eatoms[dai].normal = RDGeom::Point2D(1., 0.);
+      d_eatoms[dai].angle = -1.;
+    } else if (doneNbrs.size() == 1) {
+      auto nbid = doneNbrs.front();
+      d_eatoms[dai].nbr1 = nbid;
+      d_eatoms[dai].normal =
+          computeNormal(d_eatoms[dai].loc, d_eatoms[nbid].loc);
+    } else if (doneNbrs.size() == 2) {
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/EmbeddedFrag.cpp:421]: how does this happen?

**Review by rachelnwalker — COMMENTED** [review]
I addressed Dan's comments and added some tests. [This gist](https://gist.github.com/rachelnwalker/5c749559814512cd1e7664b77635c58d) has an overview of what some structures will look like before vs after templating. All of those templates are from coordgen, and I am definitely open to removing some of them. 

@greglandrum and @d-b-w I think this is ready for review now

File: Code/GraphMol/Depictor/Templates.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  const std::vector<RDKit::ROMol*>& getMatchingTemplates(unsigned int atom_count) {
+    return m_templates[atom_count];
+  }
+
+ private:
+  CoordinateTemplates() {
+    // load templates into m_templates map by atom count
+    std::string rdbase = getenv("RDBASE");
+    std::string fpath = rdbase + "/Code/GraphMol/Depictor/templates.smi";
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/Templates.h:52]: This is fine during development and testing, but before we merge this, the template file needs to either move to somewhere in `$RDBASE/Data` or we need to actually incorporate its contents as code. I'd lean towards direct incorporation, but I'm definitely open to discussion.

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp [hunk 7]
```diff
@@ -356,6 +430,13 @@ void EmbeddedFrag::embedFusedRings(const RDKit::VECT_INT_VECT &fusedRings) {
   RDKit::INT_VECT funion;
   RDKit::Union(fusedRings, funion);
 
+  if (useRingTemplates && fusedRings.size() > 1) {
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/EmbeddedFrag.cpp:433]: I'm wondering about the `.size()>1` test here. Couldn't there theoretically also be templates for single (non-fused) rings?

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp [hunk 8]
```diff
@@ -341,11 +308,118 @@ int EmbeddedFrag::findNeighbor(
...
+      d_eatoms[dai].angle =
+          computeAngle(d_eatoms[dai].loc, d_eatoms[nb1].loc, d_eatoms[nb2].loc);
+    } else if (doneNbrs.size() >= 3) {
+      // this is a pain - delegate it to a utility function
+      this->computeNbrsAndAng(dai, doneNbrs);
+    }
+  }
+}
+
+bool EmbeddedFrag::matchToTemplate(const RDKit::INT_VECT &ringSystemAtoms, unsigned int ring_count) {
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/EmbeddedFrag.cpp:348]: General comment that came to mind as I was reading through this function: we should be sure that we have some test molecules which match multiple templates: both in different parts of the molecule (i.e. separated by a linker) and having overlapping matches (assuming that's possible).
We also need tests looking at the interaction between templates and parts of the molecule which have fixed coordinates (like a fixed core)

File: Code/GraphMol/Depictor/testDepictor.cpp [hunk 9]
```diff
@@ -1468,6 +1469,22 @@ M  END
   }
 }
 
+
+void testUseRingTemplates() {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Depictor/testDepictor.cpp:1473]: We are, of course, gonna need a bunch more tests. :-)

File: Code/GraphMol/Depictor/RDDepictor.h [hunk 10]
```diff
@@ -77,7 +77,8 @@ RDKIT_DEPICTOR_EXPORT unsigned int compute2DCoords(
     RDKit::ROMol &mol, const RDGeom::INT_POINT2D_MAP *coordMap = nullptr,
     bool canonOrient = false, bool clearConfs = true,
     unsigned int nFlipsPerSample = 0, unsigned int nSamples = 0,
-    int sampleSeed = 0, bool permuteDeg4Nodes = false, bool forceRDKit = false);
+    int sampleSeed = 0, bool permuteDeg4Nodes = false, bool forceRDKit = false,
+    bool useRingTemplates = false);
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/RDDepictor.h:81]: We really need a version of this function which takes a params object so that it's easier to change some of these arguments.

**Review by greglandrum — COMMENTED** [review]
A bunch of comments and suggestions here.
I think this is really going to be good and very much like how well what you're doing fits into the pre-existing code base. Santosh did a really nice job way back when. :-)

File: Code/GraphMol/Depictor/Templates.h (hunk 6 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h]: I moved this to `$RDBASE/Data` for now but I can move it to the templates header if you'd prefer that. My reservation is that I envision there eventually being a couple hundred templates along with around 50 macrocycle templates, and I could see that being a little cluttered. I also use the templates file to test that all the templates are working correctly (see `testUseRingSystemTemplates` in `testDepictor.py`), so I think it'd be useful to have a file with all these templates either way.

File: Code/GraphMol/Depictor/Templates.h [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  CoordinateTemplates& operator= (const CoordinateTemplates&) = delete;
+  ~CoordinateTemplates() {
+    for (auto& [atom_cout, romols] : m_templates) {
+        for (auto& romol : romols) {
+            delete romol;
+        }
+    }
+  }
+
+  std::unordered_map<unsigned int, std::vector<RDKit::ROMol*>> m_templates;
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h]: As of now, when a template that matches a ring system is found, a pointer to the template is stored so that it can be used to get the conformer of that mol and copy over coordinates -- so I think I need to use a shared_ptr

File: Code/GraphMol/Depictor/Templates.h [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  CoordinateTemplates(const CoordinateTemplates&) = delete;
+  CoordinateTemplates& operator= (const CoordinateTemplates&) = delete;
+  ~CoordinateTemplates() {
+    for (auto& [atom_cout, romols] : m_templates) {
+        for (auto& romol : romols) {
+            delete romol;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h:67]: Ah good point!

File: Code/GraphMol/Depictor/testDepictor.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/testDepictor.cpp]: Ah ok -- I moved this test to catch_tests!

I added most of the tests in python because its a little easier to write (specifically for checking if a mol followed a template). Should I move some of these to C++ as well? Or, I could add a few more that use different structures than the python tests.

File: Code/GraphMol/Depictor/Templates.h [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  CoordinateTemplates() {
+    // load templates into m_templates map by atom count
+    std::string rdbase = getenv("RDBASE");
+    std::string fpath = rdbase + "/Code/GraphMol/Depictor/templates.smi";
+    std::ifstream cxsmiles_templates (fpath);
+    std::string smiles;
+    while(std::getline(cxsmiles_templates, smiles)) {
+        auto mol = RDKit::SmilesToMol(smiles);
+        unsigned int atom_count = mol->getNumAtoms();
+        if (m_templates.find(atom_count) == m_templates.end()) {
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h]: fixed!

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp (hunk 7 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/EmbeddedFrag.cpp:434]: Yes! I am planning to do that for macrocycles, but it will be a slightly different templating system. Those templates would be used for individual rings when there is no ring system template. From there I'd do something similar to what coordgen does -- go through possible matches for a macrocycle of a certain size and pick the layout that is best (or close to best) considering other rings and paths that come off the macrocycle.

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp (hunk 8 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/EmbeddedFrag.cpp:349]: That makes sense -- I added some python tests for this. The overlapping matches should not be possible because the matches are ring system by ring system and must be exact.

File: Code/GraphMol/Depictor/RDDepictor.h (hunk 10 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/RDDepictor.h:124]: Ok, I added a params object and named it `Compute2DCoordParameters` -- let me know if you have a better idea. I tried to make it distinguishable from `CoordGenParams`.

File: Code/GraphMol/Depictor/CMakeLists.txt [hunk 14]
```diff
@@ -14,10 +14,11 @@ endif()
 
 rdkit_library(Depictor
               RDDepictor.cpp EmbeddedFrag.cpp DepictUtils.cpp
-              LINK_LIBRARIES ${RDK_COORDGEN_LIBS} SubstructMatch MolTransforms GraphMol RDGeneral)
+              LINK_LIBRARIES ${RDK_COORDGEN_LIBS} SubstructMatch MolTransforms GraphMol RDGeneral SmilesParse)
 target_compile_definitions(Depictor PRIVATE RDKIT_DEPICTOR_BUILD)
 
 rdkit_headers(DepictUtils.h
+              Templates.h
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Depictor/CMakeLists.txt:21]: Given the current API, `Templates.h` does not need to be in rdkit_headers. Please remove it

File: Code/GraphMol/Depictor/Templates.h [hunk 15]
```diff
@@ -0,0 +1,72 @@
+//
+//  Copyright (C) 2003-2022 Greg Landrum and other RDKit contributors
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/Templates.h:2]: Copyright needs to be fixed

File: Code/GraphMol/Depictor/Wrap/rdDepictor.cpp [hunk 16]
```diff
@@ -245,7 +246,9 @@ BOOST_PYTHON_MODULE(rdDepictor) {
                  node during the sampling process \n\
      bondLength - change the default bond length for depiction \n\
      forceRDKit - use RDKit to generate coordinates even if \n\
-                  preferCoordGen is set to true\n\n\
+                  preferCoordGen is set to true\n\
+     useRingTemplates - use templates to generate coordinates of complicated\n\
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/Wrap/rdDepictor.cpp:250]:
```suggestion
     useRingTemplates - use templates to generate coordinates of complex\n\
```

File: Code/GraphMol/Depictor/testDepictor.cpp [hunk 17]
```diff
@@ -20,6 +20,7 @@
 #include "RDDepictor.h"
 #include "DepictUtils.h"
 #include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/testDepictor.cpp:23]: Why is this change here? It doesn't look like anything else in this file has ben altered.

**Review by greglandrum — CHANGES_REQUESTED** [review]
I think this is almost there. I made a few small suggestions here.

I think we'd be ok releasing an initial version of this without allowing user-editable templates, so I'm 👍 on merging this without that work done, but it would be nice if we could get in a separate PR with that functionality before the 2022.03 release.

File: Code/GraphMol/Depictor/Templates.h [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  const std::vector<std::shared_ptr<RDKit::ROMol>>& getMatchingTemplates(unsigned int atom_count) {
+    return m_templates[atom_count];
+  }
+
+ private:
+  CoordinateTemplates() {
+    // load templates into m_templates map by atom count
+    std::string rdbase = getenv("RDBASE");
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h:51]: Are you suggesting we store all templates in a single string or vector within this header, or create another header for them? I am happy to do that as long as it would still be an OK format if we had far more templates (we have 70 right now, but I could see having a couple hundred).

I've been hesitant about allowing users to submit their own templates, since they must adhere to specific requirements such as being ring systems and having standard bond lengths. Users could achieve their desired structure with alignment code, and they could add or replace templates through pull requests.

That being said, you're probably right, it would be useful (especially for organizations) to be able to define their own templates. What do you think about making it a cmake option, that way we users can't change the templates around at runtime? I'm open to other ideas as well.

File: Code/GraphMol/Depictor/Templates.h [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+      \return a reference to the singleton CoordinateTemplates
+
+      <b>Notes:</b>
+        - if the singleton CoordinateTemplates has already been instantiated
+          the singleton will be returned, otherwise the singleton will
+          be constructed.
+
+   */
+  static CoordinateTemplates& getRingSystemTemplates() {
+    static CoordinateTemplates template_mols;
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h:33]: This should be a Meyer's singleton which I believe is thread-safe https://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html

File: Code/GraphMol/Depictor/RDDepictor.cpp [hunk 20]
```diff
@@ -261,7 +263,23 @@ void embedFusedSystems(const RDKit::ROMol &mol,
...
+    if (useRingTemplates && coordMap) {
+      std::unordered_set<unsigned int> coordMapAtoms;
+      for (const auto& ring : frings) {
+        for (const auto& aid : ring) {
+          if (coordMap->find(aid) != coordMap->end()) {
+            coordMapAtoms.insert(aid);
+          }
+        }
+      }
+      allowRingTemplates = (coordMapAtoms.size() < 2);
```
[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/RDDepictor.cpp:279]: Yeah, I originally was thinking it would be OK for someone to specify the placement of a bond within the template, so two atoms would be ok -- but I think <2 is the way to go because someone could easily give a bond length that isn't compatible with the templates. I will update this.

File: Code/GraphMol/Depictor/testDepictor.cpp (hunk 17 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/testDepictor.cpp:23]: I was probably writing out SMILES to test and forgot to remove this -- I will fix that :)

File: Code/GraphMol/Depictor/Templates.h (hunk 15 above)
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h:2]: I think I just copied and pasted from another file -- what should I have here? Do I just need to update the year?

File: Code/GraphMol/Depictor/Templates.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/Depictor/Templates.h:51]: I think it's a `std::vector<string>` and can be in a `.h` or `.in` file.

[the remark this answers is not in the document]
Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Templates.h]: Ok, I added this in a new header -- lmk what you think

File: Code/GraphMol/Depictor/Wrap/testDepictor.py [hunk 21]
```diff
@@ -559,5 +559,167 @@ def testUsingCoordGenCtxtMgr(self):
...
+            mol = Chem.Mol(template)
+            mol.RemoveAllConformers()
+            rdDepictor.Compute2DCoords(mol)
+            assert not self.molMatchesTemplate(mol, template)
+
+            mol.RemoveAllConformers()
+            rdDepictor.Compute2DCoords(mol,useRingTemplates=True)
+            assert self.molMatchesTemplate(mol, template)
+
+        rdDepictor.SetPreferCoordGen(prefer_coordgen_status)
```
Comment (greglandrum) [inline Code/GraphMol/Depictor/Wrap/testDepictor.py:639]: This test no longer works since the `.smi` file has been removed.
```suggestion
```

Comment (rachelnwalker) [reply Code/GraphMol/Depictor/Wrap/testDepictor.py]: Totally forgot about that test!!

**Comment by greglandrum** [comment]
@rachelnwalker this is now failing in the SWIG wrappers. I think you don't normally look at these, so I will put together a fix and do a PR against your branch.

**Comment by greglandrum** [comment]
> @rachelnwalker this is now failing in the SWIG wrappers. I think you don't normally look at these, so I will put together a fix and do a PR against your branch.

@rachelnwalker I just made a PR against your branch which should resolve the CI problems with SWIG

**Review by greglandrum — APPROVED** [review]
LGTM!

**Comment by greglandrum** [comment]
@d-b-w I think this is ready to go. Are you happy with the most recent set of changes?

File: Code/GraphMol/Depictor/EmbeddedFrag.cpp [hunk 22]
```diff
@@ -341,11 +309,118 @@ int EmbeddedFrag::findNeighbor(
...
+  std::shared_ptr<RDKit::ROMol> template_mol(nullptr);
+  for (const auto mol : coordinate_templates.getMatchingTemplates(ringSystemAtoms.size())) {
+    // To reduce how often we have to do substructure matches, check ring info and
+    // bond count first
+    if (mol->getNumBonds() != rs_mol.getNumBonds()) {
+      continue;
+    } else if (mol->getRingInfo()->numRings() != ring_count) {
+      continue;
+    }
+    RDKit::SubstructMatchParameters params;
```
Comment (d-b-w) [inline Code/GraphMol/Depictor/EmbeddedFrag.cpp:391]: interesting, I wonder if there are other places where we really want "exact match" (graph mapping) rather than substructure search or MCS. I _presume_ that there may be some early exits possible in exact match.

**Review by d-b-w — APPROVED** [review]
I'm excited for this! Thanks, @rachelnwalker !

**Comment by greglandrum** [comment]
Thanks for this contribution @rachelnwalker! It really helps with some structures.

**Comment by ShangChien** [comment]
Any documents or examples to demonstrate the usage? 
thank you 👀
